### DISCUSSION TITLE
How to share generated sources accross modules?

### DISCUSSION THREAD

**Discussion by kincsescsaba** [sbt/sbt#7618] [Q&A]
Hey there,

I'm working with a plugin that I could use to generate sources in one of my modules; I could not configure it to generate sources in multiple modules. For this reason, I'm looking for a way to share the generated sources of my root module across most other modules. The modules should not define unnecessary classpath dependencies only to accomplish this.

Namely, this happens using the `sbt-buildinfo`, though I'd be happy to know more about such techniques independently from the related plugin.

build.sbt looks like this:
```
lazy val root = (project in file("."))
  .enablePlugins(BuildInfoPlugin)
  .settings(
    name := "myRootModule",
    buildInfoKeys ++= Seq[BuildInfoKey](
      "someJar" -> (module1 / Compile / packageBin / artifactPath).value,
      "someClassDirectory" -> (module2 / Compile / classDirectory).value
    ),
    buildInfoPackage := "buildinfo"
  )
  .aggregate(module1, module2, ....)

lazy val module1 = (project in file("module1"))
  .settings(name := "module1")

lazy val module2 = (project in file("module2"))
  .settings(name := "module2")
```

To sum up, I'd like my generated root sources to be reachable and importable in arbitrary modules.

Any ideas?

**Comment by eed3si9n** [comment]
I don't understand the question exactly. All sources must belong to a subproject, which in production will be a JAR. Are you saying you want to somehow duplicate the access to a managed source set? sbt-buildinfo is doing the right thing because the generated BuildInfo object will exist only once.

**Reply by kincsescsaba** [reply]
> Are you saying you want to somehow duplicate the access to a managed source set?

Yes, exactly. I found a workaround to do that. I set `unmanagedSources` of the submodules to point to the directory where `BuildInfo` is stored as managed source. It's ugly, but works. I needed to explicitly set `sourceManaged` using a build-level setting so that it could be reused.

My use case is rare; I need to reference, for example, packaged jars of one submodule in another. I test plugin development, where one module is a plugin to be set in the other one, and yet another module coordinates the process of testing integration of all.

I added the code below in another post.

**Comment by adpi2** [comment]
I am not sure if it is a good practice but I would try to add `root.sources` as a source generator task to `module1` and `module2`:

```scala
// use root.sources as a source generator of other projects
val commonSettings = Def.settings(
  sourceGenerators += root.sources
)

lazy val module1 = (project in file("module1"))
  .settings(
    name := "module1",
    commonSettings
  )

lazy val module2 = (project in file("module2"))
  .settings(
    name := "module2",
    commonSettings
  )
```

Hopefully sbt will be able to recognize that `module1` and  `module2` have the same source generator task and run it only once for all.

**Reply by kincsescsaba** [reply]
I'm also thinking in this direction. A similar approach worked for me; I've gone more explicit as I wanted to be sure not to mess up what the source generator plugin being used does.

Here's how it works:
```scala
val sharedGenerated = SettingKey[File]("sharedGenerated")
ThisBuild / sharedGenerated := baseDirectory.value / "sharedGenerated"

lazy val root = (project in file("."))
  .enablePlugins(BuildInfoPlugin)
  .settings(
    name := "rootModule",
    sourceManaged := (ThisBuild / sharedGenerated).value,
    buildInfoKeys ++= Seq[BuildInfoKey](
      "codeToProtectJar" -> (someModule / Compile / packageBin / artifactPath).value,
      "compilerPluginClassDirectory" -> (yetAnotherModule / Compile / classDirectory).value
    ),
    buildInfoPackage := "buildinfo"
  )
  .aggregate(someModule, yetAnotherModule, modele1, ...)

lazy val module1 = (project in file("module1"))
  .settings(
    name := "module1",
    libraryDependencies ++= Seq(
      "org.scala-lang" %% "scala3-compiler" % scalaVersion.value
    ),
    Compile / unmanagedSourceDirectories += (ThisBuild / sharedGenerated).value
  )

... more module definitions ...
```

**Reply by adpi2** [reply]
I see an issue with your approach, which is that when you run `module1 / compile` (or `module2 / compile`) you have no guarantee that the generated sources are up-to-date. Or even if you do `compile` (or `root/compile`) it is possible that the compilation of `module1` (or `module2`) starts before the sources are generated.

If you do `sourceGenerators += root.sources` it will automatically ensure that the generated sources are up-to-date before compiling.

**Reply by kincsescsaba** [reply]
You're right; my version needs to be more comprehensive.
I'm trying to implement yours, though, at this point, I'm having trouble figuring out what implicit is missing which my IDE complains about.
